core/domain/skill_services.py

```diff
@@ -277,6 +280,190 @@ def save_new_skill(committer_id, skill):
+                # skill is sufficient to apply the schema migration.
+                continue
+        return skill
+
+    except Exception as e:
+        logging.error(
+            '%s %s %s %s' % (
+                e.__class__.__name__, e, skill_id, change_list)
+        )
+        raise
+
+
+def _save_skill(committer_id, skill, commit_message, change_list):
+    """Validates a skill and commits it to persistent storage. If
+    successful, increments the version number of the incoming skill domain
+    object by 1.
+
+    Args:
+        committer_id: str. ID of the given committer.
+        skill: Skill. The skill domain object to be saved.
+        commit_message: str. The commit message.
+        change_list: list(dict). List of changes applied to a skill. Each
+            entry in change_list is a dict that represents a SkillChange.
+
+    Raises:
+        Exception: The skill model and the incoming skill domain
+            object have different version numbers.
+        Exception: Received invalid change list.
+    """
+    if not change_list:
+        raise Exception(
+            'Unexpected error: received an invalid change list when trying to '
+            'save skill %s: %s' % (skill.id, change_list))
+
+    skill_model = skill_models.SkillModel.get(
+        skill.id, strict=False)
+    if skill_model is None:
+        skill_model = skill_models.SkillModel(id=skill.id)
+    else:
+        if skill.version > skill_model.version:
+            raise Exception(
+                'Unexpected error: trying to update version %s of skill '
+                'from version %s. Please reload the page and try again.'
+                % (skill_model.version, skill.version))
+        elif skill.version < skill_model.version:
+            raise Exception(
+                'Trying to update version %s of skill from version %s, '
+                'which is too old. Please reload the page and try again.'
+                % (skill_model.version, skill.version))
+
+    skill_model.description = skill.description
+    skill_model.language_code = skill.language_code
+    skill_model.misconceptions_schema_version = (
+        skill.misconceptions_schema_version)
+    skill_model.skill_contents_schema_version = (
+        skill.skill_contents_schema_version)
+    skill_model.skill_contents = skill.skill_contents.to_dict()
+    skill_model.misconceptions = [
+        misconception.to_dict() for misconception in skill.misconceptions
+    ]
+    skill_model.commit(committer_id, commit_message, change_list)
+    memcache_services.delete(_get_skill_memcache_key(skill.id))
+    skill.version += 1
+
+
+def update_skill(committer_id, skill_id, change_list, commit_message):
+    """Updates a skill. Commits changes.
+
+    Args:
+    - committer_id: str. The id of the user who is performing the update
+        action.
+    - skill_id: str. The skill id.
+    - change_list: list of dicts, each representing a SkillChange object.
```

Comment:
Ditto (domain objects).

Also, args here don't seem to be formatted consistently with other docstrings.

Missing Raises docstring?

Reply:
Done, added the raises docstring. Will change the docstring to use domain objects.

Reply:
Done, changed to domain objects.